paimon-common/src/main/java/org/apache/paimon/fs/AsyncPositionOutputStream.java

```diff
@@ -0,0 +1,189 @@
+        this.eventQueue = new LinkedBlockingQueue<>();
+        this.exception = new AtomicReference<>();
+        this.isClosed = false;
+        this.position = 0;
+        this.future = EXECUTOR_SERVICE.submit(this::execute);
+    }
+
+    private void execute() {
+        try {
+            doWork();
+        } catch (Throwable e) {
+            exception.set(e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void doWork() throws InterruptedException, IOException {
+        try {
+            while (!isClosed) {
+                AsyncEvent event = eventQueue.poll(AWAIT_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+                if (event == null) {
+                    continue;
+                }
+                if (event instanceof EndEvent) {
+                    return;
+                }
+                if (event instanceof DataEvent) {
+                    DataEvent dataEvent = (DataEvent) event;
+                    out.write(dataEvent.data);
+                }
+                if (event instanceof FlushEvent) {
+                    out.flush();
+                    ((FlushEvent) event).latch.countDown();
+                }
+            }
+        } finally {
+            out.close();
+        }
+    }
+
+    @Override
+    public long getPos() throws IOException {
+        checkException();
+        return position;
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+        checkException();
+        position++;
+        putEvent(new DataEvent(new byte[] {(byte) b}, 0, 1));
+    }
+
+    @Override
+    public void write(byte[] b) throws IOException {
+        checkException();
+        position += b.length;
+        putEvent(new DataEvent(b, 0, b.length));
+    }
+
+    @Override
+    public void write(byte[] b, int off, int len) throws IOException {
+        checkException();
+        position += len;
+        putEvent(new DataEvent(b, off, len));
+    }
+
+    @Override
+    public void flush() throws IOException {
+        checkException();
+        FlushEvent event = new FlushEvent();
+        putEvent(event);
+        while (true) {
+            try {
+                boolean await = event.latch.await(AWAIT_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+                if (await) {
+                    return;
+                }
+                checkException();
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        checkException();
+        putEvent(new EndEvent());
+        try {
+            this.future.get();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException(e);
+        } catch (ExecutionException e) {
+            throw new RuntimeException(e);
+        } finally {
+            isClosed = true;
```

Comment:
What's the usage of this `isClosed`? The thread needs an `EndEvent` or `isClosed == true` to exit. However you use `future.get()` to wait for thread and then set `isClosed = true`, meaning that only `EndEvent` can stop the thread.